File: Coding_1103 Trees_and_Tries/최원준/최원준/3번.py
# https://leetcode.com/problems/construct-quad-tree/description/

'''
1. 아이디어 :
    리프인지 확인하는 메서드, 리프가 아닐때 반으로 자르는 메서드, 그리고 dfs를 만든다.
2. 시간복잡도 :
    O(4**n)
3. 자료구조 :
    배열
'''

"""
# Definition for a QuadTree node.
class Node:
    def __init__(self, val, isLeaf, topLeft, topRight, bottomLeft, bottomRight):
        self.val = val
        self.isLeaf = isLeaf
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight
"""

class Solution:
    def construct(self, grid: List[List[int]]) -> 'Node':

        def is_leaf(arr) -> bool:
            if not arr:
                return True
            base = arr[0][0]
            for i in range(len(arr)):
                for j in range(len(arr)):
                    if arr[i][j] != base:
                        return False
            return True

        def segment(place, arr):
            if not arr:
                return
            n = len(arr)//2
            new_arr = []
            if place == "topLeft":
                for i in range(n):
                    temp = []
                    for j in range(n):
                        temp.append(arr[i][j])
                    new_arr.append(temp.copy())

            elif place == "topRight":
                for i in range(n):
                    temp=[]
                    for j in range(n, n*2):
                        temp.append(arr[i][j])
                    new_arr.append(temp)

            elif place == "bottomLeft":
                for i in range(n, n*2):
                    temp = []
                    for j in range(n):
                        temp.append(arr[i][j])
                    new_arr.append(temp.copy())

            elif place == "bottomRight":
                for i in range(n, n*2):
                    temp=[]
                    for j in range(n, n*2):
                        temp.append(arr[i][j])
                    new_arr.append(temp.copy())

            return new_arr

        def dfs(grid):
            node = Node()
            if is_leaf(grid):
                node.isLeaf = True
                node.val = grid[0][0]
            else:
                node.isLeaf = False
                node.topLeft = dfs(segment("topLeft", grid))
                node.topRight = dfs(segment("topRight",grid))
                node.bottomLeft = dfs(segment("bottomLeft", grid))
                node.bottomRight = dfs(segment("bottomRight", grid))
            return node

        return dfs(grid)

File: Coding_231103 Trees_and_Tries/최원준/이강훈/3번.java
/*
https://leetcode.com/problems/construct-quad-tree/description/

아이디어
최소 크기인 1 까지 분할 -> 정복하면서 Leaf 여부 확인 ( 아래 4개 분할이 모두 Leaf 이면서 , 4개의 값이 모두 동일한 것이 리프 )
 -> 리프면 아래 4개 null, 아니면 유지

자료구조
변수

시간복잡도
O(N^2) // 행 * 열
*/

public class Solution {
    int[][] grid;

    public Node construct(int[][] grid) {
        this.grid = grid;

        return divide(0, 0, grid.length);
    }

    public Node divide(int x, int y, int size) {
        if (size == 1) return new Node(grid[y][x] == 1, true);

        Node node = new Node();
        node.topLeft = divide(x, y, size / 2);
        node.topRight = divide(x + size / 2, y, size / 2);
        node.bottomLeft = divide(x, y + size / 2, size / 2);
        node.bottomRight = divide(x + size / 2, y + size / 2, size / 2);

        if (node.topLeft.isLeaf && node.topRight.isLeaf && node.bottomLeft.isLeaf && node.bottomRight.isLeaf) {
            if (node.topLeft.val == node.topRight.val && node.topLeft.val == node.bottomLeft.val && node.topLeft.val == node.bottomRight.val){
                return new Node(node.topLeft.val, true);
            }
        }

        node.val = true;
        return node;
    }
}

class Node {
    public boolean val;
    public boolean isLeaf;
    public Node topLeft;
    public Node topRight;
    public Node bottomLeft;
    public Node bottomRight;


    public Node() {
        this.val = false;
        this.isLeaf = false;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }

    public Node(boolean val, boolean isLeaf) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }

    public Node(boolean val, boolean isLeaf, Node topLeft, Node topRight, Node bottomLeft, Node bottomRight) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = topLeft;
        this.topRight = topRight;
        this.bottomLeft = bottomLeft;
        this.bottomRight = bottomRight;
    }
}